### ISSUE TITLE
SC-HSM: OpenSC 0.23.0 broken on Cygwin current

### ISSUE THREAD

**Issue by clauspruefer** [OpenSC/OpenSC#2944]
### Problem Description

OpenSC 0.23.0 (current git) and 0.24.0 broken on Cygwin (Win11_64 current). Under OpenSC 0.22 everything is working fine.

### Steps to reproduce

1. Install

- Download Cygwin Win11_64 `setup-x86_64.exe` and install
- Open `Cygwin64 Terminal`
- Clone latest OpenSC git repo (`git clone https://github.com/OpenSC/OpenSC.git`)
- `./bootstrap`
- `./configure` (--strict obsolete)
- `make` (-jX)
- `make install`

2. Run pkcs-11-tool, pkcs15-tool, opensc-tool, sc-hsm-tool

- Every pkcs-11 and pkcs-15 command returns "rv = CKR_DATA_INVALID (0x20)" :(

### Proof (Compile OpenSC 0.22 with declaration-errors disabled)

- Open `Cygwin64 Terminal`
- Download OpenSC 0.22 source, depack and cd into dir
- `./bootstrap`
- `env "CFLAGS=-Wno-error=deprecated-declarations" ./configure`
- `make` (-jX)
- `make install`

=> Result: working

### Log Results

1. Without inserted card (`pkcs11-tool -L`)

Available slots:
Slot 0 (0x0): Alcor Micro USB Smart Card Reader 0
  (empty)

2. With inserted card (`pkcs11-tool -L`)

Available slots:
Slot 0 (0x0): (GetSlotInfo failed, CKR_DATA_INVALID)

Logs for 2.) attached as file.

It seems that the used smart-card (SC-HSM) has been detected (line 406) and then stuck in endless loop `Could not decode card verifiable certificate: -1401 (Invalid ASN.1 object)`. I am not too deep in pkcs15 protocol, also it could be a SC-HSM issue, but why just on CygWin, On my native Win11 (same machine) works fine.

[opensc.log](https://github.com/OpenSC/OpenSC/files/13538943/opensc.log)

### Remarks

OpenSC 0.23.0 on Win11_x64 installed is working fine (no errors with pkcs11 and pkcs15).

Also the Cygwin documentation should be updated (`--strict` compiler flag not needed anymore, also uncommenting / setting CPPFLAGS in configure.ac obslete).

**Comment by Jakuje** [comment]
We will probably need some more logs to see what is going on there and why it fails.

If you mean the wiki by the documnetation, you can fix it already using a PR to https://github.com/OpenSC/Wiki/blob/master/Compiling-on-Cygwin.md as described in the footer of wiki.

**Comment by clauspruefer** [comment]
I will provide the logs ASAP.

**Comment by clauspruefer** [comment]
Is a PR possible by just creating a new branch and pull or do i need special permission?

**Comment by clauspruefer** [comment]
Ah, ok: need to fork, that was the question.

**Comment by clauspruefer** [comment]
Ok, PR has been created, logs following.

**Comment by Jakuje** [comment]
Thanks! What about the debug logs? Do they show something useful?

**Comment by clauspruefer** [comment]
Sorry, due to usage of 0.22 on my current Win11 system (work) i have to compile on another machine. I suppose logs coming late this evening.

**Comment by clauspruefer** [comment]
Apollogizes. SC-reader in second laptop unexpected not working anymore. Very annoying. Logs following, sorry again.

**Comment by clauspruefer** [comment]
@Jakuje Attached the logs.

**Comment by Jakuje** [comment]
Thanks! @CardContact do you see something obvious in the log that could be incompatible with SC-HSM in recent OpenSC versions?
From the log, I see this is version 3.4
```
P:841; T:0x42949672976 13:34:46.036 [opensc-pkcs11] card.c:382:sc_connect_card: card info name:'SmartCard-HSM version 3.4', type:26000, flags:0x0, max_send/recv_size:238/254
```
and what is failing is in pkcs15 emulator pulling of the CVC:
```
P:841; T:0x42949672976 13:34:46.066 [opensc-pkcs11] pkcs15-sc-hsm.c:1283:sc_pkcs15emu_sc_hsm_init: called
P:841; T:0x42949672976 13:34:46.066 [opensc-pkcs11] card.c:850:sc_select_file: called; type=1, path=e82b0601040181c31f0201::
P:841; T:0x42949672976 13:34:46.067 [opensc-pkcs11] card.c:885:sc_select_file: returning with: 0 (Success)
P:841; T:0x42949672976 13:34:46.067 [opensc-pkcs11] card.c:850:sc_select_file: called; type=0, path=2f02
P:841; T:0x42949672976 13:34:46.067 [opensc-pkcs11] apdu.c:550:sc_transmit_apdu: called
P:841; T:0x42949672976 13:34:46.067 [opensc-pkcs11] card.c:471:sc_lock: called
P:841; T:0x42949672976 13:34:46.067 [opensc-pkcs11] card.c:513:sc_lock: returning with: 0 (Success)
P:841; T:0x42949672976 13:34:46.067 [opensc-pkcs11] apdu.c:515:sc_transmit: called
P:841; T:0x42949672976 13:34:46.067 [opensc-pkcs11] apdu.c:363:sc_single_transmit: called
P:841; T:0x42949672976 13:34:46.068 [opensc-pkcs11] apdu.c:367:sc_single_transmit: CLA:0, INS:A4, P1:0, P2:0, data(2) 0x7ffff79d0
P:841; T:0x42949672976 13:34:46.068 [opensc-pkcs11] reader-pcsc.c:325:pcsc_transmit: reader 'Alcor Micro USB Smart Card Reader 0'
P:841; T:0x42949672976 13:34:46.068 [opensc-pkcs11] reader-pcsc.c:326:pcsc_transmit: 
Outgoing APDU (8 bytes):
00 A4 00 00 02 2F 02 FE ...../..

P:841; T:0x42949672976 13:34:46.068 [opensc-pkcs11] reader-pcsc.c:244:pcsc_internal_transmit: called
P:841; T:0x42949672976 13:34:46.082 [opensc-pkcs11] reader-pcsc.c:335:pcsc_transmit: 
Incoming APDU (2 bytes):
67 00 g.

P:841; T:0x42949672976 13:34:46.082 [opensc-pkcs11] apdu.c:382:sc_single_transmit: returning with: 0 (Success)
P:841; T:0x42949672976 13:34:46.082 [opensc-pkcs11] apdu.c:539:sc_transmit: returning with: 0 (Success)
P:841; T:0x42949672976 13:34:46.082 [opensc-pkcs11] card.c:523:sc_unlock: called
P:841; T:0x42949672976 13:34:46.082 [opensc-pkcs11] iso7816.c:128:iso7816_check_sw: Wrong length
P:841; T:0x42949672976 13:34:46.082 [opensc-pkcs11] iso7816.c:741:iso7816_select_file: returning with: -1206 (Wrong length)
P:841; T:0x42949672976 13:34:46.083 [opensc-pkcs11] card.c:872:sc_select_file: 'SELECT' error: -1206 (Wrong length)
P:841; T:0x42949672976 13:34:46.083 [opensc-pkcs11] pkcs15-sc-hsm.c:256:read_file: Could not select EF
P:841; T:0x42949672976 13:34:46.083 [opensc-pkcs11] pkcs15-sc-hsm.c:262:read_file: Could not read EF
P:841; T:0x42949672976 13:34:46.083 [opensc-pkcs11] pkcs15-sc-hsm.c:418:sc_pkcs15emu_sc_hsm_decode_cvc: Could not decode card verifiable certificate: -1401 (Invalid ASN.1 object)
P:841; T:0x42949672976 13:34:46.083 [opensc-pkcs11] pkcs15-sc-hsm.c:1333:sc_pkcs15emu_sc_hsm_init: Could not decode EF.C_DevAut: -1401 (Invalid ASN.1 object)
```

**Comment by clauspruefer** [comment]
But as i stated out "OpenSC 0.23.0 on Windows 11_x64" works like expected with the same SC-HSM card. Maybe forgot setting an extended APDU length on CygWin build, as i remember there was a problem when the APDU size was too small with some SCs / Readers.

**Comment by Jakuje** [comment]
oh, missed that part that it works with different build. The logs looks like communication with the card is working ok. Could it be some missing dependency in the cygwin making some functionality unusable (zlib? openssl?). Can you compare output of the configure or `config.h` from our CI builds and from you cygwin?

**Comment by clauspruefer** [comment]
> oh, missed that part that it works with different build. The logs looks like communication with the card is working ok. Could it be some missing dependency in the cygwin making some functionality unusable (zlib? openssl?). Can you compare output of the configure or `config.h` from our CI builds and from you cygwin?

I will do so, i assume something is wrong with openssl changing to new api calls. APX 1-2 days...

**Comment by clauspruefer** [comment]
> > > oh, missed that part that it works with different build. The logs looks like communication with the card is working ok. Could it be some missing dependency in the cygwin making some functionality unusable (zlib? openssl?). Can you compare output of the configure or `config.h` from our CI builds and from you cygwin?
> > 
> > 
> > I will do so, i assume something is wrong with openssl changing to new api calls. APX 1-2 days...
> 
> Still in the pipe ... Too much projects in parallel ...

**Comment by clauspruefer** [comment]
> oh, missed that part that it works with different build. The logs looks like communication with the card is working ok. Could it be some missing dependency in the cygwin making some functionality unusable (zlib? openssl?). Can you compare output of the configure or `config.h` from our CI builds and from you cygwin?

Can you please state out where the CI build takes place / source / config files are located? I think i will write you some .md README files afterwards. I found the docker-container for building but no description how thw CI builds are triggered (or maybe i am blind).

**Comment by clauspruefer** [comment]
> > oh, missed that part that it works with different build. The logs looks like communication with the card is working ok. Could it be some missing dependency in the cygwin making some functionality unusable (zlib? openssl?). Can you compare output of the configure or `config.h` from our CI builds and from you cygwin?
> 
> Can you please state out where the CI build takes place / source / config files are located? I think i will write you some .md README files afterwards. I found the docker-container for building but no description how thw CI builds are triggered (or maybe i am blind).

Ha, found it, just familiar with GitLab CI, github uses `.github` folder which is at the first position where programmers never have a look at 💣

**Comment by clauspruefer** [comment]
@Jakuje Went through the CI build process (scripts), i have too less knowledge about github actions / CI. In which environment (Linux VM / Docker) the build-process is initiated?

I have some questions about the Mingw CI building.

- What exactly does the Windown Minidriver? I could not find information on github.
- Some questions about build.sh regarding Mingw integration

```bash
if [ "$1" == "mingw" -o "$1" == "mingw32" ]; then
	mkdir -p src/minidriver/CNG
	wget https://raw.githubusercontent.com/open-eid/minidriver/master/cardmod.h -O src/minidriver/CNG/cardmod.h
	if [ "$1" == "mingw" ]; then
		HOST=x86_64-w64-mingw32
	elif [ "$1" == "mingw32" ]; then
		HOST=i686-w64-mingw32
	fi
	unset CC
	unset CXX
	CFLAGS="-I$PWD/src/minidriver/CNG -Wno-error=unknown-pragmas" \
	CPPFLAGS="-DNTDDI_VERSION=0x06010000" \
	./configure --host=$HOST --with-completiondir=/tmp --disable-openssl --disable-readline --disable-zlib --enable-minidriver --enable-notify --prefix=$PWD/win32/opensc || cat config.log;
	make -j 4 V=1
	# no point in running tests on mingw
```

> Why build with openssl and zlib disabled? `./configure --host=$HOST --with-completiondir=/tmp --disable-openssl --disable-readline --disable-zlib --enable-minidriver`

```bash
if [ "$1" == "mingw" -o "$1" == "mingw32" ]; then
	# pack installed files
	wine "C:/Program Files/Inno Setup 5/ISCC.exe" win32/OpenSC.iss
fi
```

> Whats happening here?

If you could provide me the information i am able to process much faster.

**Comment by clauspruefer** [comment]
> > What exactly does the Windown Minidriver? I could not find information on github.
> 
> This is how opensc is integrated into the windows authentication system and tools. If you use pkcs11-tool, you will not get in touch with this. Generally windows provides some of its own, but allows third party to install their own, which is what OpenSC do: https://learn.microsoft.com/en-us/windows-hardware/drivers/smartcard/smart-card-minidriver-overview

Understood.

But, can you please state out where and which container (dockerfile) will be started for the build / CI process. I do not find somehow.

**Comment by Jakuje** [comment]
The mingw builds are defined here in the syntax defined for github actions, they run in `ubuntu-latest` container:

https://github.com/OpenSC/OpenSC/blob/master/.github/workflows/linux.yml#L76

But the scripts are independent from the github actions and could be executed locally with `docker`/`podman` as described here:

https://github.com/OpenSC/OpenSC/blob/master/containers/README.md

There is `opensc-mingw` Containerfile containing basically the same instructions, running the same scripts as github actions do run:

https://github.com/OpenSC/OpenSC/blob/master/containers/opensc-mingw/Containerfile

**Comment by dengert** [comment]
Can you provide opensc-debug log using the working version of OpenSC? 
Your  comment: https://github.com/OpenSC/OpenSC/issues/2944#issuecomment-1840890608 may be close to the problem. 

It could be a config option or some change in OpenSC that is being more  conservative in 0.23.0 on setting the sizes.  


In you log  at line 1698:
`card.c:382:sc_connect_card: card info name:'SmartCard-HSM version 3.4', type:26000, flags:0x0, max_send/recv_size:238/254`

The max_send/recv_size:238/254 looks way to small.

 On an older device with builtin reader on Ubuntu i see:
`card.c:382:sc_connect_card: card info name:'SmartCard-HSM version 1.2', type:26000, flags:0x0, max_send/recv_size:1232/65534` 
The card and the reader can use extended APDUs i.e. greater then 255/256 

Do you have a different reader? 

Can you try try the same code on Windows 10? (Help isolate if w10 to w11 changed something about reasder capabilities)

**Comment by clauspruefer** [comment]
> Can you provide opensc-debug log using the working version of OpenSC? Your comment: [#2944 (comment)](https://github.com/OpenSC/OpenSC/issues/2944#issuecomment-1840890608) may be close to the problem.
> 
> It could be a config option or some change in OpenSC that is being more conservative in 0.23.0 on setting the sizes.
> 
> In you log at line 1698: `card.c:382:sc_connect_card: card info name:'SmartCard-HSM version 3.4', type:26000, flags:0x0, max_send/recv_size:238/254`
> 
> The max_send/recv_size:238/254 looks way to small.
> 
> On an older device with builtin reader on Ubuntu i see: `card.c:382:sc_connect_card: card info name:'SmartCard-HSM version 1.2', type:26000, flags:0x0, max_send/recv_size:1232/65534` The card and the reader can use extended APDUs i.e. greater then 255/256
> 

I remember this is reader-driver-specific (i did this on a very old windows machine). The reader-hw settings are handled by OpenSC in Linux and CygWin similar? And if no what is the difference?

> Do you have a different reader?

Yes, imported from our office, also i am able to compile on a different machine or VM. Results following.

> 
> Can you try try the same code on Windows 10? (Help isolate if w10 to w11 changed something about reasder capabilities)

Queued.

**Comment by dengert** [comment]
Google for: "alcor micro usb smart card reader"
There appears to be some updates for this reader for Windows. 

Or try using device manager to look for new Windows 11 driver. 

ALso try setting env `OPENSC_DRIVER=sc-hsm`  to make sure OpenSC code trying to determine a card driver causes the the max_send_recv  and max_recv_size to be set with with the strange numbers before the sc-hsm driver it tested.

**Comment by clauspruefer** [comment]
> Google for: "alcor micro usb smart card reader" There appears to be some updates for this reader for Windows.
> 
> Or try using device manager to look for new Windows 11 driver.
> 
> ALso try setting env `OPENSC_DRIVER=sc-hsm` to make sure OpenSC code trying to determine a card driver causes the the max_send_recv and max_recv_size to be set with with the strange numbers before the sc-hsm driver it tested.

On Win11 with current OpenSC everything works fine on the same CygWin Machine.

**Comment by dengert** [comment]
That's good. By current, you mean 0.24.0 that was just released Dec 13, 2023?

If there is nothing else, you can close the ticket.

**Comment by clauspruefer** [comment]
> That's good. By current, you mean 0.24.0 that was just released Dec 13, 2023?
> 
> If there is nothing else, you can close the ticket.

No, i dont think so, 0.24.0 installed on Win11 -> working fine.

Compile on CygWin current on the same Win11 machine with export `OPENSC_DRIVER=sc-hsm` compiles ok.
Without `make install` `./src/tools/pkcs11-tool -O` (this should make no difference with referenced libraries) throws a Segmentation Fault. I will recompile without setting `OPENSC_DRIVER=sc-hsm`, also on different Win11 machine with different Smartcard-Reader and report. Also logs following.

**Comment by dengert** [comment]
Here is something else to try to see if problems are with  reader sizes:

[card-max_sizes.diff.txt](https://github.com/OpenSC/OpenSC/files/13688479/card-max_sizes.diff.txt)

It will list the sizes in opensc-debug log will trying to match a driver. 

OPENSC_DRIVER=sc-hsm is a run time option to tell OpenSC to  only try the sc-hsm driver.

**Comment by clauspruefer** [comment]
> Here is something else to try to see if problems are with reader sizes:
> 
> [card-max_sizes.diff.txt](https://github.com/OpenSC/OpenSC/files/13688479/card-max_sizes.diff.txt)
> 
> It will list the sizes in opensc-debug log will trying to match a driver.

I will apply patch tomorrow.

> OPENSC_DRIVER=sc-hsm is a run time option to tell OpenSC to only try the sc-hsm driver.

Good to know, this should be in every users .bashrc or similar.

**Comment by clauspruefer** [comment]
0.23.0 and 0.24.0 same behaviour with or without `OPENSC_DRIVER=sc-hsm` environment variable set. 

Also fresh Cygwin install with exact package versions from Cygwin install documentation.

```
$ pkcs11-tool -O
error: PKCS11 function C_GetSlotInfo failed: rv = CKR_DATA_INVALID (0x20)
Aborting.

```

Logs following.

**Comment by dengert** [comment]
There are too many variables in this case. Cygwin, the reader, code changes in 23. 

I tried installing Cygwin on my W11, and got as far as trying to use the Cygwin version of git which  says: "Cloning into 'OpenSC'..." then ends with no error an no clone.  

The reader appears to have have max send size of 255 and recv size of 256, which means no extended APDUs.

Commit https://github.com/OpenSC/OpenSC/commit/aadd82bb071e574fc57263a103e3bf06ebbd8de7  in card-sc-hsm.c  https://github.com/OpenSC/OpenSC/commit/aadd82bb071e574fc57263a103e3bf06ebbd8de7#diff-830e017fcc7ac0fc60f710cc6aced3b2531462147e2d4d08a1b02fbec468c812R1812-R1851
appears  set the limits  send: 255-17=238 (0xEE) and recv: 256-2= 254 (0xFE)   from lines 1821 and 1822 
"/* Assume that card supports sending with extended length APDU and without limit */"
"  card->max_recv_size = 0;" but max extended recv with extended APDU would be 65K. 

Is you w11 native or a VM?  

Can you try using the 0.24.0  windows version: https://github.com/OpenSC/OpenSC?tab=readme-ov-file#downloads i.e. eliminating the use of for Cygwin.

Do you have a Linux, Linux VM or MacOS system? 

Do you really need CygWin?

Try a reader that supports extended APDUs.

Is you card a newer version we have not seen before?
Was it initialized? 

@CardContact @istr any comments?

**Comment by clauspruefer** [comment]
> There are too many variables in this case. Cygwin, the reader, code changes in 23.
> 
> I tried installing Cygwin on my W11, and got as far as trying to use the Cygwin version of git which says: "Cloning into 'OpenSC'..." then ends with no error an no clone.
> 
> The reader appears to have have max send size of 255 and recv size of 256, which means no extended APDUs.
> 

I think this is not the case, the reader is able to handle APDU 64Kbyte (Thinkpad t490s integrated reader). On Win11 (here on the same machine also Cygwin is installed on) everything is working fine (See second Log output, but here i am not able to patch the min/max log output), also 0.22 works fine. I assume the sizes are 64k here.

> [!IMPORTANT]  
> I will compile 0.22 with the min/max output on Cygwin (@dengert: same diff patchable?) and if min/max has 64k thats the evidence.

> Commit [aadd82b](https://github.com/OpenSC/OpenSC/commit/aadd82bb071e574fc57263a103e3bf06ebbd8de7) in card-sc-hsm.c [aadd82b#diff-830e017fcc7ac0fc60f710cc6aced3b2531462147e2d4d08a1b02fbec468c812R1812-R1851](https://github.com/OpenSC/OpenSC/commit/aadd82bb071e574fc57263a103e3bf06ebbd8de7#diff-830e017fcc7ac0fc60f710cc6aced3b2531462147e2d4d08a1b02fbec468c812R1812-R1851) appears set the limits send: 255-17=238 (0xEE) and recv: 256-2= 254 (0xFE) from lines 1821 and 1822 "/* Assume that card supports sending with extended length APDU and without limit */" " card->max_recv_size = 0;" but max extended recv with extended APDU would be 65K.
> 
> Is you w11 native or a VM?

Native

> 
> Can you try using the 0.24.0 windows version: https://github.com/OpenSC/OpenSC?tab=readme-ov-file#downloads i.e. eliminating the use of for Cygwin.

0.24.0 Not working
0.23.0 Working

This seems like a change from 0.23 to 0.24 caused this. On Cagwin 0.23 also does not work, just 0.22 with old OpenSSL API calls. 

> 
> Do you have a Linux, Linux VM or MacOS system?
> 
> Do you really need CygWin?

With Cygwin you are able to use command line ssh / scp without USB-redirection to VM. This sometimes does not work expecially with exactly this "internal" laptop SC-reader ;(

> 
> Try a reader that supports extended APDUs.
> 
> Is you card a newer version we have not seen before? Was it initialized?
> 
> @CardContact @istr any comments?

**Comment by istr** [comment]
The diff is too large for me to give a quick answer to that. I see lots of changes in the pkcs15 logic between 0.22 and 0.24. And I am not acquainted in depth with the pkcs15 code.

What sounds suspicious to me is this statement:

> Cygwin uses only 1 byte (max 238/254 bytes one payload) which in my opinion indicates that setting or getting a) the readers or b) the cards properties in a wrong way. 

Given that everything works OK in Windows "native", this sounds like a specific compile-time problem with int / size_t stuff in Cygwin. I encountered something similar in a different context way back in the past with Cygwin. **But that's just a shot in the dark.**

@dengert is right to say that:

> It could be a config option or some change in OpenSC that is being more conservative in 0.23.0 on setting the sizes.

`git diff 0.22.0 0.24.0 src/libopensc/card.c` (only the relevant chunk follows)

```diff
diff --git a/src/libopensc/card.c b/src/libopensc/card.c
index f2c6ca5a..817d2b51 100644
--- a/src/libopensc/card.c
+++ b/src/libopensc/card.c
@@ -943,25 +949,57 @@ int sc_get_challenge(sc_card_t *card, u8 *rnd, size_t len)
        LOG_FUNC_RETURN(card->ctx, SC_SUCCESS);
 }
 
-int sc_read_record(sc_card_t *card, unsigned int rec_nr, u8 *buf,
-                  size_t count, unsigned long flags)
+int sc_read_record(sc_card_t *card, unsigned int rec_nr, unsigned int idx,
+                  u8 *buf ,size_t count, unsigned long flags)
 {
+       size_t max_le = sc_get_max_recv_size(card);
+       size_t todo = count;
        int r;
 
-       if (card == NULL) {
+       if (card == NULL || card->ops == NULL || buf == NULL) {
                return SC_ERROR_INVALID_ARGUMENTS;
        }
        LOG_FUNC_CALLED(card->ctx);
+       if (count == 0)
+               LOG_FUNC_RETURN(card->ctx, SC_SUCCESS);
 
        if (card->ops->read_record == NULL)
                LOG_FUNC_RETURN(card->ctx, SC_ERROR_NOT_SUPPORTED);
 
-       r = card->ops->read_record(card, rec_nr, buf, count, flags);
-       if (r == SC_SUCCESS) {
-               r = count;
+       /* lock the card now to avoid deselection of the file */
+       r = sc_lock(card);
+       LOG_TEST_RET(card->ctx, r, "sc_lock() failed");
+
+       while (todo > 0) {
+               size_t chunk = MIN(todo, max_le);
+
+               r = card->ops->read_record(card, rec_nr, idx, buf, chunk, flags);
+               if (r == 0 || r == SC_ERROR_FILE_END_REACHED)
+                       break;
+               if (r < 0 && todo != count) {
+                       /* the last command failed, but previous ones succeeded.
+                        * Let's just return what we've successfully read. */
+                       sc_log(card->ctx, "Subsequent read failed with %d, returning what was read successfully.", r);
+                       break;
+               }
+               if (r < 0) {
+                       sc_unlock(card);
+                       LOG_FUNC_RETURN(card->ctx, r);
+               }
+               if ((idx > SIZE_MAX - (size_t) r) || (size_t) r > todo) {
+                       /* `idx + r` or `todo - r` would overflow */
+                       sc_unlock(card);
+                       LOG_FUNC_RETURN(card->ctx, SC_ERROR_OFFSET_TOO_LARGE);
+               }
+
+               todo -= (size_t) r;
+               buf  += (size_t) r;
+               idx  += (size_t) r;
        }
 
-       LOG_FUNC_RETURN(card->ctx, r);
+       sc_unlock(card);
+
+       LOG_FUNC_RETURN(card->ctx, count - todo);
 }
 
 int sc_write_record(sc_card_t *card, unsigned int rec_nr, const u8 * buf,
```

Sorry that I cannot provide something better right now.

**Comment by clauspruefer** [comment]
> @dengert is right to say that:
> 
> > It could be a config option or some change in OpenSC that is being more conservative in 0.23.0 on setting the sizes.

But 0.23.0 is still working on Win64. 0.24.0 not. So here also has something changed.

**Comment by istr** [comment]
> But 0.23.0 is still working on Win64. 0.24.0 not. So here also has something changed.

Hm... :disappointed:

In that case you might need to give `git bisect` a chance to hunt down the breaking change(s).

@clauspruefer It seems that you already have the necessary build infrastructure in place. Given the large number of changes between 0.22/0.23/0.24, I fear that using git bisect is the best way to get a handle on this.

**Comment by dengert** [comment]
The other lines that do not look correct are: 
```
P:597; T:0x42949672976 21:12:00.964 [opensc-pkcs11] reader-pcsc.c:1148:detect_reader_features: Requesting reader features ... 
P:597; T:0x42949672976 21:12:00.967 [opensc-pkcs11] reader-pcsc.c:1155:detect_reader_features: Alcor Micro USB Smart Card Reader 0:SCardControl failed: 0x00000001
```
from https://github.com/OpenSC/OpenSC/blob/master/src/libopensc/reader-pcsc.c#L1148-L1156

`0x00000001` does not look like a valid return code as stated in: 
https://learn.microsoft.com/en-us/windows/win32/api/winscard/nf-winscard-scardcontrol
https://learn.microsoft.com/en-us/windows/win32/secauthn/authentication-return-values

This would point to Cygwin not handling the request for `part_10` features correctly. This combined with some OpenSC change is causing the OpenSC to support extended  APDUs for this reader.   

The code never gets to: https://github.com/OpenSC/OpenSC/blob/master/src/libopensc/reader-pcsc.c#L1281-L1291
where two readers are handled. 

Using git `bisect`  sounds like a good idea. 

Something else to try is modify the `opensc.conf` file to set reader max values as in example:
https://github.com/OpenSC/OpenSC/blob/master/etc/opensc.conf.example.in#L62-L111
This is more of a circumvention for this reader.

**Comment by clauspruefer** [comment]
@dengert @istr I also will add the log output in Cygwin/22.0 like tomorrow or in two days. Think this is very important to encircle the problem. Also i will try the `bisect` (never done before, sounds interesting) and have a bit more insight in the pkcs11 / pkcs15 transfer protocol, good night 🌃

**Comment by dengert** [comment]
I am able to reproduce the problem with Cygwin on W11 with my older 'SCM Microsystems SCx35xx v2.0 USB SC Reader 0' sc-hsm device and git clone from yesterday 

Running  under gdb with breakpoint at `detect_reader_features`
which show default is 255/256.
Then changed the reader structure to so larger 2550 and 25560
and continuing to run it lists out all the objects.   

So problem is related to not setting large enough reader max values.  
 
```
doug@DOUG22 /usr/local/bin
$ gdb --args ./pkcs11-tool -O
GNU gdb (GDB) (Cygwin 8.3.1-1) 8.3.1
Copyright (C) 2019 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Type "show copying" and "show warranty" for details.
This GDB was configured as "x86_64-pc-cygwin".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
    <http://www.gnu.org/software/gdb/documentation/>.

For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from ./pkcs11-tool...
(gdb) b detect_reader_features
Function "detect_reader_features" not defined.
Make breakpoint pending on future shared library load? (y or [n]) y
Breakpoint 1 (detect_reader_features) pending.
(gdb) run
Starting program: /usr/local/bin/pkcs11-tool -O
[New Thread 24044.0xd6c]
[New Thread 24044.0xc48]
[New Thread 24044.0x4814]
[New Thread 24044.0x5044]

Thread 1 "pkcs11-tool" hit Breakpoint 1, detect_reader_features (
    reader=reader@entry=0xa00001d20, card_handle=16861477009170104320)
    at /usr/src/opensc.b/src/libopensc/reader-pcsc.c:1136
1136    /usr/src/opensc.b/src/libopensc/reader-pcsc.c: No such file or directory.
(gdb) where
#0  detect_reader_features (reader=reader@entry=0xa00001d20,
    card_handle=16861477009170104320)
    at /usr/src/opensc.b/src/libopensc/reader-pcsc.c:1136
#1  0x00000005f8502cbd in pcsc_detect_readers (ctx=0xa00001a40)
    at /usr/src/opensc.b/src/libopensc/reader-pcsc.c:1571
#2  0x00000005f84c36d3 in sc_ctx_detect_readers (ctx=ctx@entry=0xa00001a40)
    at /usr/src/opensc.b/src/libopensc/ctx.c:767
#3  0x00000005f84c40ab in sc_context_create (
    ctx_out=ctx_out@entry=0x5f43d8140 <context>, parm=parm@entry=0x7ffffc7e0)
    at /usr/src/opensc.b/src/libopensc/ctx.c:1008
#4  0x00000005f43a2ae5 in C_Initialize (pInitArgs=<optimized out>)
    at /usr/src/opensc.b/src/pkcs11/pkcs11-global.c:349
#5  0x00000001004142e9 in main (argc=<optimized out>, argv=0xa00001850)
    at /usr/src/opensc.b/src/tools/pkcs11-tool.c:1185
(gdb) p *reader
$1 = {ctx = 0xa00001a40, driver = 0x5f8637ca0 <pcsc_drv>,
  ops = 0x5f86a4d00 <pcsc_ops>, drv_data = 0xa00001e00,
  name = 0xa00001e80 "SCM Microsystems SCx35xx v2.0 USB SC Reader 0",
  vendor = 0x0, version_major = 0 '\000', version_minor = 0 '\000', flags = 3,
  capabilities = 0, supported_protocols = 0, active_protocol = 2,
  max_send_size = 255, max_recv_size = 256, atr = {
    value = ";\376\030\000\000\201\061\376E\200\061\201THSM1s\200!@\201\a\372\000\000\000\000\000\000\000\000", len = 24}, uid = {
    value = "\000\000\000\000\000\000\000\000\000", len = 0}, atr_info = {
    hist_bytes = 0x0, hist_bytes_len = 0, Fi = 0, f = 0, Di = 0, N = 0,
    FI = 0 '\000', DI = 0 '\000'}}
(gdb) set reader->max_send_size = 2550
(gdb) set reader->max_recv_size = 25560
(gdb) p *reader
$2 = {ctx = 0xa00001a40, driver = 0x5f8637ca0 <pcsc_drv>,
  ops = 0x5f86a4d00 <pcsc_ops>, drv_data = 0xa00001e00,
  name = 0xa00001e80 "SCM Microsystems SCx35xx v2.0 USB SC Reader 0",
  vendor = 0x0, version_major = 0 '\000', version_minor = 0 '\000', flags = 3,
  capabilities = 0, supported_protocols = 0, active_protocol = 2,
  max_send_size = 2550, max_recv_size = 25560, atr = {
    value = ";\376\030\000\000\201\061\376E\200\061\201THSM1s\200!@\201\a\372\000\000\000\000\000\000\000\000", len = 24}, uid = {
    value = "\000\000\000\000\000\000\000\000\000", len = 0}, atr_info = {
    hist_bytes = 0x0, hist_bytes_len = 0, Fi = 0, f = 0, Di = 0, N = 0,
    FI = 0 '\000', DI = 0 '\000'}}
(gdb) c
Continuing.
[New Thread 24044.0x25d4]
[New Thread 24044.0x5e40]
Using slot 0 with a present token (0x0)
Certificate Object; type = X.509 cert
  label:      Certificate
  subject:    DN: C=US, ST=IL, CN=Doug Engert/emailAddress=deengert@gmail.com
  serial:     73BECD28CD5657F812F68CEBB87406D75E62A8CE
  ID:         11
Public Key Object; RSA 2048 bits
... (and the rest of objects)
```
Modifying opensc.conf did not fix problem. may need to look at this closer.)

I too have had it for the night.

**Comment by dengert** [comment]
It looks like the problem is Windows (or at least W11) does not support PCSC part 10 (or parts of it).  

If I replace `https://github.com/OpenSC/OpenSC/blob/master/src/libopensc/reader-pcsc.c#L1156` with `rcount = 0;`
(rcount should have been set when defined) I can get past the `return;` and get these log messages:

```
P:3600; T:0x42949672976 20:26:26.797 [opensc-pkcs11] ../../../opensc/src/libopensc/reader-pcsc.c:1567:pcsc_detect_readers: SCM Microsystems SCx35xx v2.0 USB SC Reader 0:SCardConnect(SHARED): 0x00000000
P:3600; T:0x42949672976 20:26:26.797 [opensc-pkcs11] ../../../opensc/src/libopensc/reader-pcsc.c:1146:detect_reader_features: called
P:3600; T:0x42949672976 20:26:26.797 [opensc-pkcs11] ../../../opensc/src/libopensc/reader-pcsc.c:1148:detect_reader_features: Requesting reader features ...
P:3600; T:0x42949672976 20:26:26.799 [opensc-pkcs11] ../../../opensc/src/libopensc/reader-pcsc.c:1303:detect_reader_features: Assuming that the reader supports sending short length APDUs only
P:3600; T:0x42949672976 20:26:26.799 [opensc-pkcs11] ../../../opensc/src/libopensc/reader-pcsc.c:1316:detect_reader_features: Assuming that the reader supports receiving short length APDUs only
```
Which are bad assumptions. 

aadd82bb071e574fc57263a103e3bf06ebbd8de7 tries to fix up two readers, but reader being used in the issue is not one of them. And two readers would not be fixe if running CygWin. The readers should be in the opensc.conf file, along with any other readers with problems. 

This also brings up the question, is this problem also in the Apvoyer release too?

And it is also looks not clear why https://github.com/OpenSC/OpenSC/blob/master/src/libopensc/reader-pcsc.c#L1370-L1376
does not find the max_send_size and max_recv_size in the opensc.conf.

Anyone else want to work on this problem?  @istr 

I really don't like CygWin.

**Comment by clauspruefer** [comment]
> I am able to reproduce the problem with Cygwin on W11 with my older 'SCM Microsystems SCx35xx v2.0 USB SC Reader 0' sc-hsm device and git clone from yesterday
> 
> Running under gdb with breakpoint at `detect_reader_features` which show default is 255/256. Then changed the reader structure to so larger 2550 and 25560 and continuing to run it lists out all the objects.
> 
> So problem is related to not setting large enough reader max values.
> 

Ah, as i expected, im sure Cygwin/22.0 will print the correct values.

@dengert But why does it work on Win11 native with 23.0 and NOT with 24.0?

**Comment by dengert** [comment]
This "/usr/local/etc/opensc.conf work for me:
```
app default {
        debug = 21;
        debug_file = /tmp/opensc-debug.txt;

        reader_driver pcsc {
                max_send_size = 65535;
                max_recv_size = 65536;
        }
}
```
Yesterday I said `opensc.conf` did not work, but I did not have spaces around "=".

Today, with only mods to opensc.conf it works. This is using  OpenSC 0.24.0 from git clone.

**But I do not get the failure 0x00000001 that you do at** https://github.com/OpenSC/OpenSC/blob/master/src/libopensc/reader-pcsc.c#L1153-L1156

Try the above opensc.conf with this mod which ignores the error:
```
diff --git a/src/libopensc/reader-pcsc.c b/src/libopensc/reader-pcsc.c
index d2dac438..9e9208b9 100644
--- a/src/libopensc/reader-pcsc.c
+++ b/src/libopensc/reader-pcsc.c
@@ -1137,7 +1137,7 @@ static void detect_reader_features(sc_reader_t *reader, SCARDHANDLE card_handle)
        sc_context_t *ctx = reader->ctx;
        struct pcsc_global_private_data *gpriv = (struct pcsc_global_private_data *) ctx->reader_drv_data;
        struct pcsc_private_data *priv = reader->drv_data;
-       DWORD rcount, i;
+       DWORD rcount = 0, i;
        u8 buf[256];
        LONG rv;
        const char *log_disabled = "but it's disabled in configuration file";
@@ -1153,7 +1153,6 @@ static void detect_reader_features(sc_reader_t *reader, SCARDHANDLE card_handle)
        rv = gpriv->SCardControl(card_handle, CM_IOCTL_GET_FEATURE_REQUEST, NULL, 0, buf, sizeof(buf), &rcount);
        if (rv != SCARD_S_SUCCESS) {
                PCSC_TRACE(reader, "SCardControl failed", rv);
-               return;
        }

        if ((rcount % sizeof(PCSC_TLV_STRUCTURE)) != 0
```

Why the difference in failure 0x0000001 and mine not failing?

- Different CygWin packages. (I set it up two days ago.) When was your updated?

- I am testing with older SC-HSM device

- Difference in Windows.  I have:
```
Device name	DOUG22
Processor	12th Gen Intel(R) Core(TM) i5-12600K   3.69 GHz
Installed RAM	32.0 GB (31.7 GB usable)
Device ID	A087B189-ECEE-4B65-8347-DB565AAB1C64
Product ID	00330-71290-00002-AAOEM
System type	64-bit operating system, x64-based processor
Pen and touch	No pen or touch input is available for this display
Edition	Windows 11 Pro
Version	22H2
Installed on	‎10/‎8/‎2022
OS build	22621.2861
Experience	Windows Feature Experience Pack 1000.22681.1000.0
```

Why native works and CygWin fails? Don't know, other then CygWin maybe internals maybe a problem.  But it looks like OpenSC is calling the Windows pcsc DLL directly.

**Comment by frankmorgner** [comment]
We could use the libccid data on the readers [dwMaxCCIDMessageLength](https://ccid.apdu.fr/ccid/dwMaxCCIDMessageLength.html) to add more special cases:

https://ccid.apdu.fr/ccid/section.html

not sure how to tackle this in a more generic way. maybe pyscard can be used for more debugging on the Windows side. Maybe also @LudovicRousseau has insights about how to get property data on Windows (or its limitations)

**Comment by frankmorgner** [comment]
Am I right in assuming that the use case of the OP involves `writing` to the card? Meaning: Extended length support is required for the transaction in question.

In cardos, we simply *assume* that extended length is supported in this case:
https://github.com/OpenSC/OpenSC/blob/b0d0707b15aee2c06f2c5922f27812b0ca9ca005/src/libopensc/card-cardos.c#L330-L340

Maybe this would also be an option for sc-hsm

**Comment by dengert** [comment]
@clauspruefer  I have been looking CygWin and dont have an answer to why
`rv = gpriv->SCardControl(card_handle, CM_IOCTL_GET_FEATURE_REQUEST, NULL, 0, buf, sizeof(buf), &rcount);` is failing. 


I am waiting for you to try git bisect.

**Comment by LudovicRousseau** [comment]
You can try to use [get_PCSC_Part10_features.py](https://github.com/LudovicRousseau/PCSC-contrib/blob/master/get_PCSC_Part10_features.py) 

With my Gemalto PC Twin Reader reader I get:
```
± ./get_PCSC_Part10_features.py 
Reader: Gemalto PC Twin Reader
CM_IOCTL_GET_FEATURE_REQUEST results:
 FEATURE_GET_TLV_PROPERTIES: 0x42330012
  Supports Secure PIN Entry (SPE): no
FEATURE_GET_TLV_PROPERTIES results:
[1, 2, 0, 0, 3, 1, 0, 8, 18, 71, 101, 109, 84, 119, 105, 110, 45, 86, 50, 46, 48, 48, 45, 71, 70, 48, 51, 9, 1, 0, 11, 2, 230, 8, 12, 2, 55, 52, 10, 4, 0, 0, 1, 0]
[3, 1, 0, 8, 18, 71, 101, 109, 84, 119, 105, 110, 45, 86, 50, 46, 48, 48, 45, 71, 70, 48, 51, 9, 1, 0, 11, 2, 230, 8, 12, 2, 55, 52, 10, 4, 0, 0, 1, 0]
[8, 18, 71, 101, 109, 84, 119, 105, 110, 45, 86, 50, 46, 48, 48, 45, 71, 70, 48, 51, 9, 1, 0, 11, 2, 230, 8, 12, 2, 55, 52, 10, 4, 0, 0, 1, 0]
[9, 1, 0, 11, 2, 230, 8, 12, 2, 55, 52, 10, 4, 0, 0, 1, 0]
[11, 2, 230, 8, 12, 2, 55, 52, 10, 4, 0, 0, 1, 0]
[12, 2, 55, 52, 10, 4, 0, 0, 1, 0]
[10, 4, 0, 0, 1, 0]
 raw: [1, 2, 0, 0, 3, 1, 0, 8, 18, 71, 101, 109, 84, 119, 105, 110, 45, 86, 50, 46, 48, 48, 45, 71, 70, 48, 51, 9, 1, 0, 11, 2, 230, 8, 12, 2, 55, 52, 10, 4, 0, 0, 1, 0]
 PCSCv2_PART10_PROPERTY_wLcdLayout: 0 or 0x0000
 PCSCv2_PART10_PROPERTY_bTimeOut2: 0 or 0x0000
 PCSCv2_PART10_PROPERTY_sFirmwareID: GemTwin-V2.00-GF03
 PCSCv2_PART10_PROPERTY_bPPDUSupport: 0 or 0x0000
 PCSCv2_PART10_PROPERTY_wIdVendor: 2278 or 0x08E6
 PCSCv2_PART10_PROPERTY_wIdProduct: 13367 or 0x3437
 PCSCv2_PART10_PROPERTY_dwMaxAPDUDataSize: 65536 or 0x10000
```
The important line here is: ` PCSCv2_PART10_PROPERTY_dwMaxAPDUDataSize: 65536 or 0x10000`


I guess that the Windows CCID driver does NOT support `FEATURE_GET_TLV_PROPERTIES`. But I have not tested.
Maybe non-Microsoft drivers may support `FEATURE_GET_TLV_PROPERTIES`.

**Comment by dengert** [comment]
@LudovicRousseau  See below you may be correct. 

The Windows SDK `winsmcrd.h` has:
```
#define SCARD_CTL_CODE(code)        CTL_CODE(FILE_DEVICE_SMARTCARD, \
                                            (code), \
                                            METHOD_BUFFERED, \
                                            FILE_ANY_ACCESS)
...
METHOD_BUFFERED = 0, FILE_ANY_ACCESS = 0

#define IOCTL_SMARTCARD_GET_FEATURE_REQUEST SCARD_CTL_CODE(3400)
```
and "Interoperability Specification for ICCs and Personal Computer Systems
Part 10 IFDs with Secure PIN Entry Capabilities" from 2012 Says:

> Implementation Note for Windows:
Device_Success is implemented as SCARD_S_SUCCESS
Under Windows the control code must be defined as follows:
#define CM_IOCTL_GET_FEATURE_REQUEST SCARD_CTL_CODE(3400)

The above implies the Windows pcsc should work. 

VirtualBox linux guests can read the FEATURE_GET_TLV_PROPERTIES. 

But the "Microsoft Usbccid Smartcard Reader (WUDF)" is dated 6/21/2006. Maybe they never did implemented it. 

https://answers.microsoft.com/en-us/windows/forum/all/download-windows-11-pro-microsoft-usbccid/0098e568-a196-4cff-9b31-dd13465b1777?page=1 with 173 "I have the same problem" and  13,089 views implies they never updated the driver, despite what the header files say and Microsoft are co-authors of the part10 document!

OpenSC  may have worked without the FEATURE_GET_TLV_PROPERTIES because other OpenSC code was able get past the  255/256 limits. but some recent changes have forced the opensc to use 255/256

Waiting for @istr to try the `git biscet` to find out which commit fails.

**Comment by istr** [comment]
@dengert Just to clarify

> Waiting for @istr to try the git biscet to find out which commit fails.

I will not dig into that myself, because I do not have (easy) access to a Win / Cygwin build env right now.
My proposal was a hint for @clauspruefer what he could try to hunt down the breaking change.

**Comment by dengert** [comment]
Sorry, @clauspruefer said he would try the git bisect. 

Look at comments in #2964 It looks like Windows PCSC and CCID driver do not support this call:
https://github.com/OpenSC/OpenSC/blob/master/src/libopensc/reader-pcsc.c#L1153 maybe returning `rv = 0` or `rv = 1` and since `rcount` is not initialized, it may not change and have other problems.   So none of: 
https://github.com/OpenSC/OpenSC/blob/master/src/libopensc/reader-pcsc.c#L1165-L1185
are ever set and the `dwMaxAPDUDataSize` is never read. 

CygWin calls the Windows winscard.dll, and the problem is not just CygWin but the build by "continuous-integration/appveyor/pr" hasthe same problem.  

Also looks like another mod set the default max_send_size=255 and max_recv=256 first.  So even the check for "REINER SCT cyberJack" and "Secure Flash Card" may not be done either. We don't know what the max sizes on Windows is for any reader.

**Comment by LudovicRousseau** [comment]
To use [get_PCSC_Part10_features.py](https://github.com/LudovicRousseau/PCSC-contrib/blob/master/get_PCSC_Part10_features.py) on Windows you need to replace `SCARD_SHARE_DIRECT` by `SCARD_SHARE_SHARED` **and** insert a card in the reader.

With the default/Windows CCID driver I get nothing.

With the Gemalto driver and the same Gemalto PC Twin Reader I get:
```
>python get_PCSC_Part10_features.py
Reader: Gemplus USB Smart Card Reader 0
CM_IOCTL_GET_FEATURE_REQUEST results:
 FEATURE_GET_TLV_PROPERTIES: 0x312140
 FEATURE_CCID_ESC_COMMAND: 0x312144
  Supports Secure PIN Entry (SPE): no
FEATURE_GET_TLV_PROPERTIES results:
 raw: [10, 4, 0, 0, 0, 0, 11, 2, 230, 8, 12, 2, 55, 52, 8, 18, 71, 101, 109, 84, 119, 105, 110, 45, 86, 50, 46, 48, 48, 45, 71, 70, 48, 51, 9, 1, 0]
 PCSCv2_PART10_PROPERTY_dwMaxAPDUDataSize: 0 or 0x0000
 PCSCv2_PART10_PROPERTY_wIdVendor: 2278 or 0x08E6
 PCSCv2_PART10_PROPERTY_wIdProduct: 13367 or 0x3437
 PCSCv2_PART10_PROPERTY_sFirmwareID: GemTwin-V2.00-GF03
 PCSCv2_PART10_PROPERTY_bPPDUSupport: 0 or 0x0000
````

Here we have `PCSCv2_PART10_PROPERTY_dwMaxAPDUDataSize: 0 or 0x0000` which indicates "0: short APDU only." http://pcscworkgroup.com/Download/Specifications/pcsc10_v2.02.09.pdf

**Comment by dengert** [comment]
Started bisect on CyGwin  (then went shopping) using these commands:

```
$ git bisect start HEAD 5497519ea6b4af596628f8f8f2f904bacaa3148f --  # tag 0.23.0 
$ git bisect run /tmp/test.sh
```
```
#!/bin/sh
cd /usr/src/opensc
git clean -xf
./bootstrap || exit 125
./configure  || exit 125
make -j 3 || exit 125
make install || exit 125
# the test program
OPENSC_DRIVER=sc-hsm /usr/local/bin/pkcs11-tool -O
```
and as expected got this hit:

```
aadd82bb071e574fc57263a103e3bf06ebbd8de7 is the first bad commit
commit aadd82bb071e574fc57263a103e3bf06ebbd8de7
Author: Ingo Struck (git commits) <gitlab@ingostruck.de>
Date:   Sat Jan 21 22:15:10 2023 +0100

    Handle reader limits for SC Card unwrap operations

    Fixes #2514

 src/libopensc/card-sc-hsm.c            | 181 ++++++++++++++++++---------------
 src/libopensc/reader-pcsc.c            |  91 ++++++++++-------
 src/tests/fuzzing/fuzz_pkcs15_decode.c |   3 +-
 src/tests/fuzzing/fuzz_pkcs15_encode.c |   2 +-
 4 files changed, 159 insertions(+), 118 deletions(-)
bisect found first bad commit
```

@LudovicRousseau points out the Gemalto driver driver works, but not the Microsoft CCID driver from 2006 which is being used in my case: 

` rv = gpriv->SCardControl(card_handle, CM_IOCTL_GET_FEATURE_REQUEST, NULL, 0, buf, sizeof(buf), &rcount);`
returns `rv = 0;  rcount=0;`
and  `reader` already has defaults set for   `max_send_size = 255, max_recv_size = 256`
```
 {ctx = 0xa00001b60, driver = 0x5f8631c80 <pcsc_drv>,
  ops = 0x5f869cd00 <pcsc_ops>, drv_data = 0xa00001f00,
  name = 0xa00001f80 "SCM Microsystems SCx35xx v2.0 USB SC Reader 0",
  vendor = 0x0, version_major = 0 '\000', version_minor = 0 '\000', flags = 3,
  capabilities = 0, supported_protocols = 0, active_protocol = 2,
  max_send_size = 255, max_recv_size = 256, atr = {
    value = ";\376\030\000\000\201\061\376E\200\061\201THSM1s\200!@\201\a\372\000\000\000\000\000\000\000\000", len = 24}, uid = {
    value = "\000\000\000\000\000\000\000\000\000", len = 0}, atr_info = {
    hist_bytes = 0x0, hist_bytes_len = 0, Fi = 0, f = 0, Di = 0, N = 0,
    FI = 0 '\000', DI = 0 '\000'}}
```
So none of the features are returned,  and this like is never run:
`int max_data = part10_detect_max_data(reader, card_handle);`
and the defaults are left at 255/256.

**Proposed solution**

We have no control over the reader,  its version, the windows version or device driver and the type of card. Also do not have control over these on any OS.   
Add a flags to reader indicating where the sizes came from. 
part10_detect_max_data set the reader sizes  sizes set from default or from opensc.conf. 

i.e. consider the max sizes as a best guess.

It would then allow a card driver or change the sizes if it knew they were too small, for example 
if the driver requires extended APDU, because the card does not support chaining and/or get response. 

If the driver does support chaining and/or get response but the driver (or apdu.c routines) could try extended and if it works, set new max sizes and fall back to 255/256 as last resort. 

Also aadd82bb071e574fc57263a103e3bf06ebbd8de7 is forcing the max size to be smaller because of a limit on read/write binary, but this then applies to all operations. Other card drivers handle this by temporarily changine the sizes when doing operations the requiring a smaller size.

**Comment by clauspruefer** [comment]
> Sorry, @clauspruefer said he would try the git bisect.
> 

Sorry guys, had to take a break. Project should have been closed on last sunday... But still ongoing... No sleep @ weekend.

**Comment by dengert** [comment]
> Sorry guys, had to take a break. 

No problem, it turned out running the git bisect with a script was easier then I thought it would be. The bisect took about 8 builds to find the commit but did it unattended.  

 >  Project should have been closed on last sunday...
Not sure what you mean.   What you have found is a problem not just CygWin with a specific card, but a problem in native windows build and maybe other cards too.  

Its the holidays and I too will have little time to work on this but there is a circumvention as listed in #2964.

**Comment by clauspruefer** [comment]
> > Sorry guys, had to take a break.
> 
> No problem, it turned out running the git bisect with a script was easier then I thought it would be. The bisect took about 8 builds to find the commit but did it unattended.
> 
> > Project should have been closed on last sunday...
> > Not sure what you mean.   What you have found is a problem not just CygWin with a specific card, but a problem in native windows build and maybe other cards too.

I mean my own project im currently working on :smile:
> 
> Its the holidays and I too will have little time to work on this but there is a circumvention as listed above. Also see #2964.

Saw it already will try ASAP :runner:

> [!TIP]
> Just a small proposal of handling all upcoming issues with the misbehaving readers.
>
> Why not adding / setting the readers making problems to an exlusion list (config / structs / matrix), just setting the 65k manually for just those readers. No source code change for dependend problems (including testing) necessary. Furthermore if people reporting problems with a special reader: just add this reader to the config.

**Comment by dengert** [comment]
>Just a small proposal of handling all upcoming issues with the misbehaving readers.

The problem is not necessarily the reader. The main problem apprears to be the windows CCID device driver from 2006 does needs updating. When the same reader is used on non windows machine it works fine. It works fine from a virtual guest machine with Windows host.  (VirtualBox may be talking to the USB device and bypassing the windows CCID  reader driver.)

The [Part 10 IFDs with Secure PIN Entry Capabilities](https://pcscworkgroup.com/Download/Specifications/pcsc10_v2.02.09.pdf) added support for `dwMaxAPDUDataSize` in November 2012.  Unfortunately, many of my readers predate the initial release of Part_10. 

Using Windows Device Manager, with your reader and card inserted can you look at "smart card reader", and look properties for the driver being used and its date. Form the initial  post, the name is: "Alcor Micro USB Smart Card Reader" 

The problem for the `sc-hsm` may be the commit in question sets the sizes less then the default for **all operations** for the card instead of only when doing specific ADPUs. 

The circumvention was provide relief for users with this problem while continuing to use existing software. OpenSC only does releases about once a year.  Most users only have one card. A more robust solution needs to be developed to work with future OS, readers and cards hopefully without having to list individual readers. One problem is across OSes and Virtual software vendors where the reader name seen in the guest in not the same as seen from the host.

**Comment by clauspruefer** [comment]
![image](https://github.com/OpenSC/OpenSC/assets/17313789/0eded45c-c4e6-4529-8b4b-ba4b2c5f4647)

**Comment by LudovicRousseau** [comment]
@dengert most of PCSC v2 part 10 is done by the driver, not the reader. It could work even with very old readers.

For example `dwMaxAPDUDataSize` is implemented [here](https://github.com/LudovicRousseau/CCID/blob/master/src/ifdhandler.c#L1838) in my CCID driver.

**Comment by clauspruefer** [comment]
> @dengert most of PCSC v2 part 10 is done by the driver, not the reader. It could work even with very old readers.
> 
> For example `dwMaxAPDUDataSize` is implemented [here](https://github.com/LudovicRousseau/CCID/blob/master/src/ifdhandler.c#L1838) in my CCID driver.

And why does it work with the same Win CCID driver on 23.0 and not on 24.0?

**Comment by dengert** [comment]
> For example dwMaxAPDUDataSize is implemented [here](https://github.com/LudovicRousseau/CCID/blob/master/src/ifdhandler.c#L1838) in my CCID driver.

The problem is only on Windows. So your driver or pcsc-lite  is not used. So the Windows problems may be in their PC/SC or some CCID driver. 

And why does it work with the same Win CCID driver on 23.0 and not on 24.0?

On Windows ` rv = gpriv->SCardControl(card_handle, CM_IOCTL_GET_FEATURE_REQUEST, NULL, 0, buf, sizeof(buf), &rcount);`
when I run it, returns rv = 0 and rcount = 0, so no feature are returned and some of the code following is executed which changes max sizes to 238/254. 

The git bisect shows [aadd82bb](https://github.com/OpenSC/OpenSC/commit/aadd82bb071e574fc57263a103e3bf06ebbd8de7) causes the problem. 
It changes the max sizes in in reader-pcsc.c for sc-hsm to the 238/254 which means if short APDUs are used, the card may complain as seen in some debug logs that P1-P2 are in error, but I think it because the card wants to return more then 256 bytes of data using get-response, but the the host says it only wants 254. Le is usually `00` which says return  256 bytes or less and the host will then request more data using get response. If host only asks for 254, the card assumes host will not  use get response. and returns an error.  (I can not prove this at this time.) 

See:  https://github.com/OpenSC/OpenSC/issues/2944#issuecomment-1859304468 in one of the logs is:
`reader-pcsc.c:1155:detect_reader_features: Alcor Micro USB Smart Card Reader 0:SCardControl failed: 0x00000001`
This then take a different path. and returns immediately.  [aadd82bb](https://github.com/OpenSC/OpenSC/commit/aadd82bb071e574fc57263a103e3bf06ebbd8de7) also has mods to sc-hsm that changed max sizes to less then 255/256 which can cause the same problem.

I can not do any real testing of this until after the holidays. 

@clauspruefer if you wanted to use git  to revert  aadd82bb  and run you tests that would help show the problem is in   aadd82bb.   But   aadd82bb also fixes issues symmetric key encryption, which needs to be addressed by temporarily setting the max sizes but only for some operations but  not for **all** operations.

**Comment by dengert** [comment]
Found some time. Works with CygWin with this patch:  [sc-hsm-cygwin.diff.txt](https://github.com/OpenSC/OpenSC/files/13796437/sc-hsm-cygwin.diff.txt)
when running:  `pkcs11-tool --test --login`

card-sc-hsm.c  now set the max_send_size to at least 256, so the card can return data, When it was 254, the card would return errors, as suggested in https://github.com/OpenSC/OpenSC/issues/2944#issuecomment-1868327428

The change to reader-pcsc.c was to make sure some of the tests where still run even if 0x0000001 was returned. 
The card driver will still use extended APDUs in some cases. for example sign. as the reader still supports them and can handle 64K packets.  

I would like @istr and @clauspruefer to give this a try. 

I still don't understand all of what  https://github.com/OpenSC/OpenSC/commit/aadd82bb071e574fc57263a103e3bf06ebbd8de7 is trying to do.

**Comment by clauspruefer** [comment]
I will give a try and report back.

**Comment by clauspruefer** [comment]
# Correct Version 0.24

```bash
$ pkcs11-tool -I
Cryptoki version 3.0
Manufacturer     OpenSC Project
Library          OpenSC smartcard framework (ver 0.24)
Using slot 0 with a present token (0x0)
```

# Object List ok

- one x.509 cert (for use with MIcrosoft CrytoAPI
- one public SSH key (private hidden, no login given)

```
$ pkcs11-tool -O
Using slot 0 with a present token (0x0)
Certificate Object; type = X.509 cert
  label:      CP-RSA4096-sshKey
  subject:    DN: C=DE, ST=BW, L=Mannheim, O=clickIT, OU=Security, CN=c.pruefer.ssh/emailAddress=pruefer@webcodex.de
  serial:     8B65287CC95FDC3E
  ID:         01
Public Key Object; RSA 4096 bits
  label:      CP-RSA4096-sshKey
  ID:         01
  Usage:      encrypt, verify, verifyRecover
  Access:     local
Profile object 340144
  profile_id:          CKP_PUBLIC_CERTIFICATES_TOKEN (4)
```

@dengert Close issue or wait for istr reply?

**Comment by dengert** [comment]
The patch just proves one problem is making the `max_recv_size` to small and many commands would fail because the card would treat the 254 as an error. The card sill assumes it can use extended APDU and it does and the reader supports it. The problem on Windows is it can return  `dwMaxAPDUDataSize` and the code then tries to force the 255/256 limits and further limited max_recv_size to 254 in `sm-hsm.c` 

In other words, on windows OpenSC can not detect the reader sizes and should not set artificial limits. In the sc-hsm case the card drive  to the most part  assumes the reader will support extended APDU large enough for its needs. Other card drivers may  fail back to using the 255/256 and use command chaining and get response to handle larger outgoing  commands and incoming data.  

Wait for @istr as there are still unanswered questions as to how to fix the symmetric key issue with read/write binary operations with odd number APDU instructions. 

@clauspruefer 
If you run with an opensc debug log and use pkcs11-tool --test --login, can you look for a simple sign operation. 
With a 2048 bit RSA key  which I have on my older token the easiest one would look like this:

```
Outgoing APDU (265 bytes):
80 68 02 20 00 01 00 00 01 FF FF FF FF FF FF FF .h. ............
FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF ................
FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF ................
...
4A 74 73 A1 C0 0B C1 02 00 
```
where `80 68 02 20` is the APDU,  Lc `00`  means extended APDU with  length `01 00` for 256 bytes and the last two bytes are the extended Le `02 00` the max size to return. 

With your 4096 bit key, the outgoing would be 521 bytes with Lc length of `02 00` and Le would still be `02 00` as it looks like the code just has a fixed size buffer and the card will return the actual size.

The  response  would look like:
```
Incoming APDU (258 bytes):
53 B5 AD 07 8B 71 8C F8 D1 A4 D6 57 6C 29 45 34 S....q.....Wl)E4
7B 1D 87 A5 82 67 E0 AC 6B 31 F0 07 AC F0 D4 8F {....g..k1......
...
90 00
```
But  for 514 bytes different bytes, ending in `90 00`

**Comment by clauspruefer** [comment]
> If you run with an opensc debug log and use pkcs11-tool --test --login, can you look for a simple sign operation.

On the run. I think it might also be helpful to check if i can gen a new key with `sc-hsm-tool` on another card, so i will try this first.

**Comment by dengert** [comment]
@istr  and  @CardContact

Thanks for running the test with 4K key. So we are close. 

The main problem is Windows does not support ` gpriv->SCardControl(card_handle, CM_IOCTL_GET_FEATURE_REQUEST, NULL, 0, buf, sizeof(buf), &rcount);` and combined with aadd82bb07 this causes CygWin commands certutil.exe /v /SCinfo  to fail. 

```
C:\Users\doug>certutil.exe /v -SCinfo
The Microsoft Smart Card Resource Manager is running.
Current reader/card status:
Readers: 1
  0: SCM Microsystems SCx35xx v2.0 USB SC Reader 0
--- Reader: SCM Microsystems SCx35xx v2.0 USB SC Reader 0
--- Status: SCARD_STATE_PRESENT | SCARD_STATE_UNPOWERED
--- Status: The card is available for use.
---   Card: SmartCard-HSM
---    ATR:
        3b fe 18 00 00 81 31 fe  45 80 31 81 54 48 53 4d   ;.....1.E.1.THSM
        31 73 80 21 40 81 07 fa                            1s.!@...


=======================================================
Analyzing card in reader: SCM Microsystems SCx35xx v2.0 USB SC Reader 0
P:21636; T:26224 2024-01-04 10:59:35.388 [cardmod] PKCS#15 init failed.
P:21636; T:26224 2024-01-04 10:59:35.640 [cardmod] PKCS#15 init failed.
P:21636; T:26224 2024-01-04 10:59:35.892 [cardmod] PKCS#15 init failed.
...
```
Changes in  aadd82bb07 in `reader-pcsc.c` cause the defaults of the card max_send_size and max_recv_size to be set to zero.

I have some questions about aadd82bb07 and  existing `card-sc-hsm.c` code:

- OpenSC has routines  `sc_get_max_send_size(card)` and  `sc_get_max_recv_size(card)` why are these not used? They test if card driver has set  SC_CARD_CAP_APDU_EXT.
- Why the  `card->max_recv_size = card->reader->max_recv_size - 2;` ? Three card drivers do this and one says for SW1 and SW2. But that is not the case, the send and receive sizes are size of data sent and received. `reader-pcsc.c` takes care of this. 
- `// 17 byte header and TLV because of odd ins in UPDATE BINARY` and  `card->max_send_size = card->reader->max_send_size - 17;` Why the `- 17`?  `sc_hsm_write_ef` takes care of  the TLV header bytes. And I assume the extra 10 bytes is for SM?    
- In any case the `card->max_recv_size` should not be less then 256 as the card appears to have problems if it is less the 256. 

@clauspruefer   Thanks for running the test with 4K key. So we are close.

**Comment by frankmorgner** [comment]
- According to the SC-HSM user manual (I only had v2.4 at hand), SELECT doesn't seem to support a variable value for Le and MUST be 0x00 (or absent if P2 == 0x0C). 
- Le also MUST be exactly 0x00 for Generate Session PIN (INS=0x5A), General Authenticate (INS=86), Import DKEK (INS=52), Manage PKA (INS=54)
- Interestingly, the Le for Enumerate Objects (sc_hsm_list_files), Generate Asymmetric Key Pair (sc_hsm_generate_keypair), Wrap Key (sc_hsm_wrap_key) similarly MUST be 0x000000 (i.e. extended length)

All of the above commands are explicitly setting the Le to 0 (and encode extended lenght if needed) except the select command, which instead calls the iso7816 driver for selecting. A quick way for fixing this, would be to force Le = 0 in the select:

```diff
diff --git a/src/libopensc/card-sc-hsm.c b/src/libopensc/card-sc-hsm.c
index e68975077..a236a2cbc 100644
--- a/src/libopensc/card-sc-hsm.c
+++ b/src/libopensc/card-sc-hsm.c
@@ -151,6 +151,8 @@ static int sc_hsm_select_file_ex(sc_card_t *card,
 	sc_hsm_private_data_t *priv = (sc_hsm_private_data_t *) card->drv_data;
 	sc_file_t *file = NULL;
 	sc_path_t cpath;
+	size_t card_max_recv_size = card->max_recv_size;
+	size_t reader_max_recv_size = card->reader->max_recv_size;
 
 	if (file_out == NULL) {				// Versions before 0.16 of the SmartCard-HSM do not support P2='0C'
 		rv = sc_hsm_select_file_ex(card, in_path, forceselect, &file);
@@ -184,7 +186,11 @@ static int sc_hsm_select_file_ex(sc_card_t *card,
 				&& in_path->aid.len == sc_hsm_aid.len
 				&& !memcmp(in_path->aid.value, sc_hsm_aid.value, sc_hsm_aid.len))) {
 		if (!priv || (priv->dffcp == NULL) || forceselect) {
+			/* Force use of Le = 0x00 in iso7816_select_file as required by SC-HSM */
+			card->max_recv_size = card->reader->max_recv_size = 256;
 			rv = (*iso_ops->select_file)(card, in_path, file_out);
+			card->max_recv_size = card_max_recv_size;
+			card->reader->max_recv_size = reader_max_recv_size;
 			LOG_TEST_RET(card->ctx, rv, "Could not select SmartCard-HSM application");
 
 			if (priv) {
@@ -213,14 +219,24 @@ static int sc_hsm_select_file_ex(sc_card_t *card,
 			*file_out = file;
 			return SC_SUCCESS;
 		} else {
+			/* Force use of Le = 0x00 in iso7816_select_file as required by SC-HSM */
+			card->max_recv_size = card->reader->max_recv_size = 256;
 			sc_path_t truncated;
 			memcpy(&truncated, in_path, sizeof truncated);
 			truncated.len = in_path->len - 2;
 			memcpy(truncated.value, in_path->value+2, truncated.len);
-			return (*iso_ops->select_file)(card, &truncated, file_out);
+			rv = (*iso_ops->select_file)(card, &truncated, file_out);
+			card->max_recv_size = card_max_recv_size;
+			card->reader->max_recv_size = reader_max_recv_size;
+			return rv;
 		}
 	}
-	return (*iso_ops->select_file)(card, in_path, file_out);
+	/* Force use of Le = 0x00 in iso7816_select_file as required by SC-HSM */
+	card->max_recv_size = card->reader->max_recv_size = 256;
+	rv = (*iso_ops->select_file)(card, in_path, file_out);
+	card->max_recv_size = card_max_recv_size;
+	card->reader->max_recv_size = reader_max_recv_size;
+	return rv;
 }
```
An alternative would be to explicitly implement the APDU encoding for SELECT in sc-hsm instead of forwarding to iso7816.

I hope the above suggestion fixes the problem.

**Comment by dengert** [comment]
Looks reasonable. I don't have the manual. Does the SC-HSM manual talk about Le = 00 with short APDUs and use of get-response? In which case OpenSC apdu.c would do the get response.  Or does SC-HSM require the use of an extended reader?
The reader in question does support  extended APDUs, as the circumvention of setting 65535/65535 works.

The main problem is OpenSC has no way on Windows to query the reader for is max sizes and the card-sc-hsm.c sets max sizes less then 255/256

**Comment by frankmorgner** [comment]
The manual does not mention GET RESPONSE and also doesn't mention command chaining for anything else than device authentication. If I remember correctly, @CardContact already confirmed that there is no support for this and that extended length APDUs are indeed required in cases with big chunks of data.

**Comment by frankmorgner** [comment]
I've created a PR, so that you can pick binaries from the CI pipeline for testing https://github.com/OpenSC/OpenSC/pull/2978

**Comment by frankmorgner** [comment]
32 bit binaries are available here
https://ci.appveyor.com/project/frankmorgner/opensc/builds/48931227/job/xts017729w423c66/artifacts

64 bit binaries here
https://ci.appveyor.com/project/frankmorgner/opensc/builds/48931227/job/7v9lyv4ml9p1ftdk/artifacts

**Comment by frankmorgner** [comment]
@clauspruefer would you mind testing the suggested fix, please?

**Comment by frankmorgner** [comment]
With "the last patch" you mean https://github.com/OpenSC/OpenSC/pull/2978?

**Comment by clauspruefer** [comment]
I currently do not have access to Github, it was a small patch in a 
reply to Issue #294. I will recheck when i have access to Github (this 
evening).

Regards,
Claus

**Comment by clauspruefer** [comment]
The diff attached to the comment from dengert: https://github.com/OpenSC/OpenSC/issues/2944#issuecomment-1872383403

**Comment by frankmorgner** [comment]
then please Test the above pull request or binaries 
https://github.com/OpenSC/OpenSC/issues/2944#issuecomment-1888871069

This should be the correct fix

**Comment by dengert** [comment]
The  real problem is `part10_find_property_by_tag`  `PCSCv2_PART10_PROPERTY_dwMaxAPDUDataSize` is not from the reader hardware but from PCSC, CCID and/or IFD software. Microsoft has not implemented it.  It is only in PCSClite. 
The code in `reader-pcsc.c` does not handle the failure correctly and dies not check for a few known readers.   

As @LudovicRousseau  said in: https://github.com/OpenSC/OpenSC/issues/2944#issuecomment-1868075374 PCSClite implements `PCSCv2_PART10_PROPERTY_dwMaxAPDUDataSize`

It is not just a Cygwin problem, it is a problem for OpenSC code run on windows both which use `reader-pcsc.c` which will fail on Windows. It may affect other card drivers too. 
If we really wanted it to work, we would add other readers to the list in `reader-pcsc.c`.

See: https://github.com/OpenSC/OpenSC/files/13796437/sc-hsm-cygwin.diff.txt which handles the failure better.

**Comment by frankmorgner** [comment]
Sorry, but I disagree. The real problem (tm) seems to be that SELECT is called with Le!=0, which is *caused* by not having PCSCv2_PART10_PROPERTY_dwMaxAPDUDataSize at hand. We should be able to cope with PCSCv2_PART10_PROPERTY_dwMaxAPDUDataSize unset or even set incorrectly on all PC/SC platforms, but SELECT should always be issued with Le=0, as required by the SC-HSM specification.

**Comment by dengert** [comment]
My point in https://github.com/OpenSC/OpenSC/files/13796437/sc-hsm-cygwin.diff.txt is that  when 
`SCardControl(card_handle, CM_IOCTL_GET_FEATURE_REQUEST,...` fails at:
https://github.com/OpenSC/OpenSC/blob/master/src/libopensc/reader-pcsc.c#L1153

the 178 lines of code: 
https://github.com/OpenSC/OpenSC/blob/master/src/libopensc/reader-pcsc.c#L1158-L1337
are not executed, which includes:

- code to test for specific readers
- pin pad info (probably why pin pad readers don't work in Windows) 
- `part10_detect_pace_capabilities(reader, card_handle);`
- `max_send_size = max_recv_size = part10_detect_max_data(reader, card_handle);`
- other code to `set max_send_size` and `max_recv_size` which may end in "Assuming that the reader supports sending short length APDUs only"

This results in a card driver like `card-sc-hsm.c` expects the reader in the token to support extended APDUs that the reader does not.  

This lack of the features from PCSClite in the Windows implementation could be addressed if OpenSC took some of the changes from PCSClite to fill in info for readers. Or at least look in `opensc.conf`

This problem may be the same problem as in #3004  and other card drivers may have similar problems. 

(I actually found that building OpenSC from github using  Cygwin was easy and could debug using `gdb`)

**Comment by frankmorgner** [comment]
Thats indeed another Problem for which ive created a seperate issue

**Comment by Jakuje** [comment]
What is the status of this issue? Was it fixed with  #3006 ? Is it fixed with #2978 ?

**Comment by dengert** [comment]
I don't know.

**Comment by frankmorgner** [comment]
both patches are needed to reliably fix this in all situations
